### ISSUE TITLE
Are "codec" stats per transceiver or per transport?

### ISSUE THREAD

**Issue by jan-ivar** [w3c/webrtc-stats#614]
TL;DR: The spec states codec stats are per transport, not per m-line, which doesn't match Chrome. But whaddabout sdpFmtpLine?

["codec"](https://w3c.github.io/webrtc-stats/#dom-rtccodecstats) _"is created when a codec type is registered for an **RTP transport**. It may be **referenced by multiple RTP streams in media sections** using that transport, but similar codecs in different transports have different RTCCodecStats objects."_

While _"RTP transport"_ isn't clearly defined, there's a `transportId` member, so I'd assume this refers to that `"transport"` which [says](https://w3c.github.io/webrtc-stats/#dom-rtccodecstats-transportid): _"The unique identifier of the transport on which this codec is being used, which can be used to look up the corresponding RTCTransportStats object."_

From [testing](https://jsfiddle.net/jib1/5v4qdr79/) I see Chrome creating a full set of `"codec"` entries (46) for _each_ transceiver (10 transceivers = 460 "codec" entries), even though there's only 1 `"transport"` (because BUNDLE). I read the spec as sayng that 10 transceivers should produce 46 "codec" instances (in Chrome), not 460.

But I'm not sure what this means for the sdpFmtpLine, since I believe `a=fmtp` may vary per m-line (but how often does it)?

One one hand this looks like a spec bug. On the other, returning 460+ JS objects seems like a waste and potential scaling issue.

cc @docfaraday @pehrsons

**Comment by Pehrsons** [comment]
Another corner case is when an fmtp line differs in the local and remote sdps, so the encoder and decoder would be configured differently. If this difference in sdpFmtpLine should be surfaced through stats, the codec stats in this case end up being per rtp-stream rather than per transceiver. Related to #616.

Should codec stats be coalesced up to transport-level while all parameters match? That's certainly an option that avoids lots of waste in most cases.

**Comment by alvestrand** [comment]
460 codec entries sounds like a bug to me.
The intent when writing the "codec" stats was that one would be able to know the precise configuration for each payload type; the way RTP works, there can't be two different codec configurations for the same payload type on the same transport in the same direction, so it only makes sense to have max 1 codec entry for each PT.

**Comment by jan-ivar** [comment]
> so it only makes sense to have max 1 codec entry for each PT.

...per direction in some cases, but yes this makes sense to me.

**Comment by henbos** [comment]
460??? Yeahhhh, I'd say that's a bug

**Comment by vr000m** [comment]
In callstats.js, we just filtered out the duplicates, hence, coalescing does make sense.

**Comment by fippo** [comment]
Given the recent motion to reduce codec stats greatly can we revisit this one? With @henbos latest proposal to only produce codecs that are in use this brings down the number from 4 to 2 but deduping seems more effort than these two objects are worth

**Comment by henbos** [comment]
You're referring to the "per direction" note specifically, can you file a separate issue for that? The issue that codecs are per-transport and not per-transceiver still holds.

**Comment by fippo** [comment]
I am refering to the coalescing note that was added in #617 which is what completed this issue

**Comment by alvestrand** [comment]
sadly, the PT can refer to different configs in the two directions of a transport.
I think it makes sense to think of the fmtp as "this is what I have promised to handle if you send it to me" - thus, the two ends can give different promises, but making two different promises at the same end (such as would happen if fmtp differed between two PT descriptions in the same bundle) makes absolutely no sense to me.
